### DISCUSSION TITLE
OR-Tools SCIP Stop Solver on User Action

### DISCUSSION THREAD

**Discussion by GhayourAli** [google/or-tools#3628] [Linear Solver questions]
I am using .Net wrapper of OR Tools to solve a Mixed Integer Programming problem using SCIP solver. Is there a way to stop the solver based on user action e.g. user wants to stop the solver and check the solution value reached till that time? I found one solution using CPSolver but I don't think it will work with SCIP. I tried using Solver::InterruptSolver method but it did not seemed to have stopped the solver.  Also, is there a way to know how many solutions solver has checked so far?

**Comment by lperron** [comment]
1) InterruptSolver() should work.
2) no.

If your problem mixed ? (having continuous variables).
If not, CP-SAT is a better non-mixed IP solver that SCIP. And it supports solution callback and StopSearch().

**Reply by GhayourAli** [reply]
i tried InterruptSolver, but the solver kept running until the specified time limit. Also, when I checked the objective value right after interrupt, it returned 0 as objective value. Does solver stop immediately after InterruptSolver?

Yes, the problem is mixed. Most of the variables are integers but only a few variables are continuous.

**Reply by GhayourAli** [reply]
Following is the code snippet,

```
Task.Run(()=> {
      Thread.Sleep(10000);
       solver.InterruptSolver();
       Print(solver.objective.value());
}
solver.Solve();
Print(solver.objective.value());
```

**Comment by lperron** [comment]
https://github.com/google/or-tools/blob/stable/ortools/linear_solver/scip_interface.cc#L116

and the method is "InterruptSolve()", not "InterruptSolver()".

**Reply by GhayourAli** [reply]
Apologies for the mistake in the snippet. I am in fact using InterruptSolve() but it seems like it does not stop the solver and it keeps running. Below is the exact code snippet. After 1 minute, I call InterruptSolve() method and it returns true (Objective value is printed as 0 at this point).. But after 2 mins (time limit set for solver), solver.Solve(parameter) returns Feasible and correct objective value is printed.

```cs
    //This is how I have created the solver before adding objective and constraints
    //var solver = Solver.CreateSolver("SCIP");
    //solver.SetTimeLimit(120000);
        
    //This is the solve method where I am trying to interrupt.
    MPSolverParameters parameters = new MPSolverParameters();
    parameters.SetDoubleParam(MPSolverParameters.DoubleParam.RELATIVE_MIP_GAP, 0.03);

    Task.Run(() =>
    {
        Thread.Sleep(60000);
        var hasStopped = solver.InterruptSolve();
        Console.WriteLine(solver.Objective().Value());
    });
    var resultStatus = solver.Solve(parameters);
    Console.WriteLine(solver.Objective().Value());
```

**Comment by GhayourAli** [comment]
@lperron Solver::InterruptSolve() seems to work however if I set the solver number of threads to 8, then it does not work and solver keeps solving even after InterruptSolve() has been called. Is this the expected behavior or is this some kind of an issue that needs to be fixed?

**Comment by lperron** [comment]
do not use parallelism with SCIP. It is buggy, unreliable. It leaks
memory...
And the gain is minimal.
Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00



Le mar. 10 janv. 2023 à 09:17, GhayourAli ***@***.***> a
écrit :

**Reply by GhayourAli** [reply]
Ohk, thanks for the heads up.
